src/sentry/profiles/flamegraph.py

```diff
@@ -612,3 +612,63 @@ def get_spans_based_candidates(self, query: str | None, limit: int) -> EAPRespon
                 auto_fields=True,
             ),
         )
+
+
+def split_datetime_range_exponential(
+    start_datetime: datetime,
+    end_datetime: datetime,
+    initial_chunk_delta: timedelta,
+    max_delta: timedelta,
+) -> Iterator[tuple[datetime, datetime]]:
+    """
+    Splits a datetime range into exponentially increasing chunks, yielded by a generator.
+
+    The duration of each chunk doubles from the previous one until it reaches the
+    max_delta, at which point the chunk size remains constant.
+
+    Args:
+        start_datetime (datetime): The start of the datetime range.
+        end_datetime (datetime): The end of the datetime range.
+        initial_chunk_delta (timedelta): The duration of the first chunk.
+        max_delta (timedelta): The maximum duration for any chunk.
+
+    Yields:
+        tuple: A tuple representing a datetime chunk (start_of_chunk, end_of_chunk).
+
+    Raises:
+        TypeError: If args are not the correct datetime/timedelta objects.
+        ValueError: If datetimes are invalid, or deltas are not positive,
+                    or initial_chunk_delta > max_delta.
+    """
+    if not all(isinstance(dt, datetime) for dt in [start_datetime, end_datetime]):
+        raise TypeError("start_datetime and end_datetime must be datetime objects.")
+
+    if not all(isinstance(td, timedelta) for td in [initial_chunk_delta, max_delta]):
+        raise TypeError("initial_chunk_delta and max_delta must be timedelta objects.")
```

Comment:
These checks seem redundant when we have type hints.

Reply:
type hints are more of a "suggestion" in Python.

`mypy` can help with the static analysis but what happens at runtime it's out of mypy scope.

Better safe than sorry!